File: C/Problem0021.c
/**
 * Problem0021 - Amicable numbers
 * 
 * https://github.com/agileshaw/Project-Euler
 **/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int divisorSum(int num)
{
	int sum = 1, i;
	
	for (i = 2; i < sqrt(num+1); i++) {
		if (num % i == 0){
			if (num/i == i)
				sum += i;
			else
				sum += i + num/i;
		}
	}

	return sum;
}

int main()
{
	int i, j, size = 10000, result = 0;
	int* num = calloc(size, sizeof(int));

	if (num == NULL) {
		printf("Error allocating memory\n");
		exit(1);
	}
	
	for (i = 2; i < size; i++)
		num[i] = divisorSum(i);

	for (i = 1; i < size; i++) {
		j = num[i];
		if (j < size && num[j] == i && j != i)
			result += i;
	}

	printf("The sum of amicable numbers is: %d\n", result);
	free(num);

	return 0;
}

File: Python/Problem0021.py
# Problem0021 - Amicable numbers
#
# https: // github.com/agileshaw/Project-Euler

import math

def divisorSum(num):
	dsum = 1
	for i in range(2, int(math.sqrt(num+1))):
		if num % i == 0:
			if num//i == i:
				dsum += i
			else:
				dsum += i + num//i
	return dsum

if __name__ == "__main__":
	result = 0
	size = 10000
	num = [0] * size
	for i in range(2, size):
		num[i] = divisorSum(i)
	for i in range(1, size):
		j = num[i]
		if j < size and num[j] == i and i != j:
			result += i
	print("The sum of amicable numbers is: %d" % result)
